src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,161 @@
+        if (taskNo < 0 || taskNo >= Task.getTaskCount()) {
+            throw new DukeException("OOPS!! Please enter a valid task number to mark!");
+        }
+        return new MarkAsDoneCommand(taskNo);
+    }
+
+    public static Command parseMarkAsUndone(String desc) throws DukeException{
+        Scanner scanner = new Scanner(desc);
+        if (!scanner.hasNextInt()) {
+            throw new DukeException("OOPS!! Please enter a valid task number to unmark!");
+        }
+        int taskNo = scanner.nextInt();
+        if (taskNo < 0 || taskNo >= Task.getTaskCount()) {
+            throw new DukeException("OOPS!! Please enter a valid task number to unmark!");
+        }
+        return new MarkAsUndoneCommand(taskNo);
+
+    }
+
+    public static Command parseDelete(String desc) throws DukeException {
+        Scanner scanner = new Scanner(desc);
+        if (!scanner.hasNextInt()) {
+            throw new DukeException("OOPS!! Please enter a valid task number to delete!");
+        }
+        int taskNo = scanner.nextInt();
+        if (taskNo < 0 || taskNo >= Task.getTaskCount()) {
+            throw new DukeException("OOPS!! Please enter a valid task number to delete!");
+        }
+        return new DeleteCommand(taskNo);
+    }
+
+    public static Command parseFind(String desc) {
+        return new FindCommand(desc);
+    }
+
+    public static String parseDate(String date) {
+        LocalDate localDate = LocalDate.parse(date.trim());
+        String formatDate = localDate.format(DateTimeFormatter.ofPattern("dd MMM yyyy"));
+        return formatDate;
+    }
+
+    public static Command parse(String rawCommand) throws DukeException {
+        Scanner sc = new Scanner(rawCommand);
+        String first = sc.next();
+        String desc = "";
+        if (sc.hasNext()) {
+            desc = sc.nextLine().trim();
+        }
+
+        switch (first) {
+            case ("bye"):
+                return new ByeCommand();
+
+            case ("list"):
+                return new ListCommand();
+
+            case ("todo"):
+                return Parser.parseTodo(desc);
+
+            case ("deadline"):
+                return Parser.parseDeadline(desc);
+
+            case ("event"):
+                return Parser.parseEvent(desc);
+
+            case ("mark"):
+                return Parser.parseMarkAsDone(desc);
+
+            case ("unmark"):
+                return Parser.parseMarkAsUndone(desc);
+
+            case ("delete"):
+                return Parser.parseDelete(desc);
+
+            case ("find") :
+                return Parser.parseFind(desc);
+
+            default:
+                throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
+
+    public static Command parseFileLine(String desc) throws DukeException {
+        String[] words = desc.split("\\s\\|\\s");
+        int size = words.length;
+        if (size == 1) return new NullCommand();
```

Comment:
I noticed that your NullCommand() throws an exception that is not caught, and returns nothing.

Perhaps it will be more suitable to create a Duke exception that handles a command with no description, and remove the NullCommand class?